### PR TITLE
Improved HEC error messages

### PR DESCRIPTION
**Pull request by sbylica-splunk**
**Description:**  Improved error messages returned by HEC

**Testing:** Unit testing

### CODE REVIEW

File: .chloggen/improved_hec_error_msgs.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: splunkhecexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Added messages returned by Splunk to error displayed in logs when there is some issue with exporting events with HEC
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [ADDON-68220]
```
Comment (atoulme) [inline .chloggen/improved_hec_error_msgs.yaml:13]: typically should be a github issue on this repo or the PR number:
```suggestion
issues: [34339]
```

File: internal/splunk/httprequest.go [hunk 2]
```diff
@@ -29,6 +31,20 @@ func HandleHTTPCode(resp *http.Response) error {
...
+		var jsonResponse map[string]any
+		bodyString, _ := io.ReadAll(resp.Body)
+		unmarshalError := json.Unmarshal(bodyString, &jsonResponse)
+		
+		if unmarshalError == nil {
+			responseErrorValue, ok := jsonResponse["text"]
+			if ok {
+				err = multierr.Append(err, fmt.Errorf("Reason: %q", responseErrorValue))
+			}
+		}
```
Comment (atoulme) [inline internal/splunk/httprequest.go:45]: else?

File: internal/splunk/httprequest.go [hunk 3]
```diff
@@ -29,6 +31,20 @@ func HandleHTTPCode(resp *http.Response) error {
...
+	// Check if there is any error text returned by Splunk that we can append to error message
+	if resp.Body != nil {
+		var jsonResponse map[string]any
+		bodyString, _ := io.ReadAll(resp.Body)
+		unmarshalError := json.Unmarshal(bodyString, &jsonResponse)
+		
+		if unmarshalError == nil {
+			responseErrorValue, ok := jsonResponse["text"]
+			if ok {
+				err = multierr.Append(err, fmt.Errorf("Reason: %q", responseErrorValue))
```
Comment (atoulme) [inline internal/splunk/httprequest.go:43]: iirc errors in go must start lower-case

File: internal/splunk/httprequest.go [hunk 4]
```diff
@@ -29,6 +31,20 @@ func HandleHTTPCode(resp *http.Response) error {
 		resp.StatusCode,
 		http.StatusText(resp.StatusCode))
 
+	// Check if there is any error text returned by Splunk that we can append to error message
```
Comment (atoulme) [inline internal/splunk/httprequest.go:34]: please add a test?

Comment (sbylica-splunk) [reply internal/splunk/httprequest.go]: Yup, done

File: internal/splunk/httprequest.go (hunk 2 above)
Comment (atoulme) [reply internal/splunk/httprequest.go:47]: else nothing then? We don't care if we can't read the JSON?

**Comment by crobert-1** [comment]
From failing CI/CD action:
```
Generated code is out of date, please run "make generate" and commit the changes in this PR.
```
Below lint failure can be fixed by `make gofmt`, I believe.
```
splunk/httprequest.go:6: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
import (
splunk/httprequest.go:7: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"fmt"
splunk/httprequest.go:12: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"io"
	"encoding/json"
make[2]: *** [../../Makefile.Common:199: lint] Error 1
```

**Comment by sbylica-splunk** [comment]
Updated the pr

**Comment by crobert-1** [comment]
_Looks like more tests need to be updated to properly handle this change._

File: internal/splunk/httprequest.go [hunk 5]
```diff
@@ -29,6 +33,23 @@ func HandleHTTPCode(resp *http.Response) error {
 		resp.StatusCode,
 		http.StatusText(resp.StatusCode))
 
+	// Check if there is any error text returned by Splunk that we can append to the error message
+	if resp.Body != nil {
+		var jsonResponse map[string]any
+		bodyString, _ := io.ReadAll(resp.Body)
+		resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyString))
```
Comment (sbylica-splunk) [inline internal/splunk/httprequest.go:40]: Added this line so the body can be reused later

**Pull request closed without merging**